### PR TITLE
Increase performance of the UDP client

### PR DESCRIPTION
**Pull request by nkatsaros**
I made the UDP client only resolve and dial once, when it's created. This does two things: it makes Write significantly faster, and it pushes name resolution errors to when you first create the client. 

I also added a buffer size knob the new UDP config struct I created. This is very important, as before the client was likely broken unless you were using it on your own machine. MTUs will generally not be ~65000, they will be more like 1500 or 500 and this needs to be tuneable if people plan on using this client over the internet. Unfortunately, UDPConn.SetReadBuffer sets the OS read buffer size and will not change the maximum size datagram you can receive so I removed it.

Lastly, in the UDP service I moved buffer creation outside of the loop to save allocations.

This is just an idea. I don't know if you'd want to commit to an extra function on the Client interface or the UDPConfig struct, but I believe they are necessary.

### CODE REVIEW

**Comment by otoolep** [comment]
Thanks @nkatsaros -- your changes generally make sense to me.

File: services/udp/service.go [hunk 1]
```diff
@@ -137,8 +137,8 @@ func (s *Service) serve() {
 	defer s.wg.Done()
 
 	s.batcher.Start()
+	buf := make([]byte, UDPBufferSize)
```
Comment (otoolep) [inline services/udp/service.go]: How much testing did you perform of this change?

**Comment by otoolep** [comment]
Have you signed the CLA?

https://influxdb.com/community/cla.html

File: client/v2/client.go [hunk 2]
```diff
@@ -86,12 +98,28 @@ func NewClient(conf Config) Client {
...
 // NewUDPClient returns a client interface for writing to an InfluxDB UDP
-// service address, specified by addr. addr should be of the form
-// "host:port" or "[ipv6-host%zone]:port". Write will fail if addr is invalid.
-func NewUDPClient(addr string) Client {
-	return &udpclient{
-		addr: addr,
+// service from the given config.
+func NewUDPClient(conf UDPConfig) (Client, error) {
+	var udpAddr *net.UDPAddr
+	udpAddr, err := net.ResolveUDPAddr("udp", conf.Addr)
```
Comment (otoolep) [inline client/v2/client.go]: Yes, moving this to the New() (or perhaps an Open()) call is absolutely the right thing to do. We should have caught this earlier.

File: services/udp/service.go (hunk 1 above)
Comment (nkatsaros) [inline services/udp/service.go]: I'll be honest, I didn't. I don't know why it would cause an issue because the buffer is only used for the duration of that loop, and the size of the read is used to slice it. This can probably be left out because @sparrc looks like he's changing it in his udp-perf branch.

**Comment by nkatsaros** [comment]
Signed the CLA. I've contributed on telegraf before but I don't remember if I signed it then.

File: services/udp/service.go (hunk 1 above)
Comment (otoolep) [inline services/udp/service.go]: The reason I asked is we tried this change before and it result in corrupted buffers. We meant to look into why, and forgot. It does seem like it a valid change, but it didn't work for us.

Can you revert this change? I really like your client changes, and we want those. We can come back to this issue.

Comment (sparrc) [inline services/udp/service.go]: yes, please just revert this part of the change, it'll be changing in my udp-perf branch anyways :+1:

File: client/v2/client.go [hunk 3]
```diff
@@ -61,6 +70,9 @@ type Client interface {
 	// Query makes an InfluxDB Query on the database. This will fail if using
 	// the UDP client.
 	Query(q Query) (*Response, error)
+
+	// Close releases any resources a Client may be using.
+	Close() error
```
Comment (sparrc) [inline client/v2/client.go]: you will need to write

**Comment by sparrc** [comment]
@nkatsaros should the `MaximumUDPBufferSize` default to something lower? based on what you said, maybe it should be 1500?

File: client/v2/client.go (hunk 2 above)
Comment (sparrc) [inline client/v2/client.go]: good! :+1:

**Comment by nkatsaros** [comment]
1500-28 (ip and udp header size) would be a decent default for a network that doesn't have jumbo frames enabled.

**Comment by nkatsaros** [comment]
Pushed a new commit with the buffer change in the UDP service removed.

File: client/v2/client.go [hunk 4]
```diff
@@ -447,3 +460,13 @@ func (c *client) Query(q Query) (*Response, error) {
 	}
 	return &response, nil
 }
+
+// Close releases the udpclient's resources.
+func (uc *udpclient) Close() error {
+	return uc.conn.Close()
+}
+
+// Close releases the client's resources.
```
Comment (otoolep) [inline client/v2/client.go]: For readability's sake, this be grouped with the `Client`.

File: client/v2/client.go (hunk 3 above)
Comment (otoolep) [inline client/v2/client.go]: The only thing about this is that we're adding a `Close()` without an `Open()`. If we had an `Open()` it would be there we could do the name resolution. I think symmetry is important in this interface.

File: client/v2/client.go (hunk 4 above)
Comment (nkatsaros) [inline client/v2/client.go]: Agreed. I was mirroring the existing style.

File: client/v2/client.go (hunk 3 above)
Comment (nkatsaros) [inline client/v2/client.go]: I've always paired New/Close or Open/Close together depending on context. The same thing is kind of done in the [stdlib](https://golang.org/pkg/net/rpc/#NewClient). I would be against adding Open because it adds complexity to the API and isn't necessary for the http client.

Comment (otoolep) [inline client/v2/client.go]: In this case I think `Open` is better. Is a UDPClient opened if `err` is non-nil on its instantiation? Is it OK to call `Close` on a non-nil UDP client? Etc, etc.

Thoughts?

Comment (nkatsaros) [inline client/v2/client.go]: In my experience, the contract on New or Open has always been that if `err` is non-nil then you must not call Close, and if it is nil then you must (should) call Close in the future.

Maybe the real solution here is to have the user create their own `net.Conn` and pass it in. That way they can tweak whatever knobs they want and the burden of connection lifecycle is passed on to them.

**Comment by sparrc** [comment]
Re Buffer Size. Maybe we should we call that constant `MTUSize` so that people definitely understand what they are adjusting. Then we set the buffer to `MTUSize - 28`

File: client/v2/client.go (hunk 3 above)
Comment (sparrc) [inline client/v2/client.go]: I'd rather not make users create their own `net.Conn`, seems overly complicated. IMO it's okay to have a `Close()` function without an `Open()`. In the `net` library itself the `Close()` function is paired with the `DialUDP()` function.

This also means that we need `NewUDPClient` to return an error, correct?

**Comment by nkatsaros** [comment]
Agreed, but do you then determine if the connection is ipv4 (20 byte header) or ipv6 (40 byte header) and subtract the correct value?

I would also keep in mind for your docs that UDP packets over 1024 bytes are slowed down considerably on Windows unless you tune the [FastSendDatagramThreshold registry key](https://www.microsoft.com/windows/windowsmedia/howto/articles/optimize_web.aspx).

**Comment by sparrc** [comment]
FWIW, we might just want to play it safe and go all the way down to 512 as the default. http://stackoverflow.com/questions/1098897/what-is-the-largest-safe-udp-packet-size-on-the-internet

I'm also thinking that calling it `MTUSize` would be a mistake, since it appears that sending your MTU size might not be the best idea.

Safety is good, so now I'm thinking we call it `UDPPacketSize` and default to `512`, thoughts?

**Comment by otoolep** [comment]
I am OK with reducing it the level that is supported by everyone. As long as it is configurable, we should be OK. 

That said, sending UDP over the Internet tells me people don't really care much about their data.

File: client/v2/client.go (hunk 3 above)
Comment (otoolep) [inline client/v2/client.go]: Yeah, let's not force them pass in  a `net.Conn`. It's too much. I'm OK with not having an `Open()`, it's not a blocker. The changes here are a big improvement regardless.

**Comment by otoolep** [comment]
+1, though we may want to reduce the buffer size as discussed.

**Comment by otoolep** [comment]
Yeah, let's reduce it.

https://github.com/influxdb/influxdb/issues/4656#issuecomment-154077943

**Comment by nkatsaros** [comment]
Renamed the config option to PayloadSize because it's more descriptive. Removed UDP from the option name because it's redundant. Changed the default payload size to 512.

Any problems with these changes?

**Comment by otoolep** [comment]
Looks great.

**Comment by otoolep** [comment]
Thanks again for your contribution @nkatsaros

**Comment by sparrc** [comment]
@otoolep LGTM, do you min dif I merge it?

**Comment by otoolep** [comment]
do-it

**Comment by sparrc** [comment]
wow, nice
